threedworld-mit/tdw
# Stop Replicant animation

# Test

1. Run the latest release standalone build from master. In `tdw`, `master` branch, run this:

```python
from tdw.controller import Controller
from tdw.tdw_utils import TDWUtils
from tdw.add_ons.replicant import Replicant
from tdw.replicant.arm import Arm
from tdw.replicant.action_status import ActionStatus
from tdw.add_ons.third_person_camera import ThirdPersonCamera


c = Controller()
replicant = Replicant()
camera = ThirdPersonCamera(position={"x": -0.5, "y": 1.175, "z": 3},
                           look_at=replicant.replicant_id,
                           avatar_id="a")
c.add_ons.append(replicant)
c.add_ons.append(camera)
c.communicate([TDWUtils.create_empty_room(12, 12),
               {"$type": "set_target_framerate",
                "framerate": 30}])
replicant.animate(animation="limping")
for i in range(200):
    c.communicate([])
c.communicate([])
replicant.reach_for(target={"x": 0.2, "y": 0.5, "z": 0.3}, arm=Arm.right, absolute=False,
                    scale_duration=False, duration=1)
while replicant.action.status == ActionStatus.ongoing:
    c.communicate([])
c.communicate([])
```

Expected result: The Replicant interrupts its animation and reaches for a target.
Actual result: The Replicant continues its animation

2. Run [this test build](https://www.dropbox.com/s/8bi0ymse677bxc1/TDW_Windows_replicant_stop_animation.zip?dl=0). In `tdw`, on the test branch `replicant_stop_animation`, run the same controller.

Result: The Replicant interrupts its animation and reaches for a target.

# Changelog

### Command API

#### New Commands

| Command                    | Description                          |
| -------------------------- | ------------------------------------ |
| `stop_replicant_animation` | Stop an ongoing Replicant animation. |

### `tdw` module

- Fixed: If a Replicant starts a new action while an animation is playing (i.e. `animate(animation)`, `move_by(distance)`, or `move_to(target)`), the animation doesn't stop.
